### PR TITLE
[Don't merge yet] Shutdown support (alternative try at #1684)

### PR DESCRIPTION
**Pull request by bendk**
In the review for #4790, @mhammond suggested trying a one-shot approach at interruption.  This is my try at it.

Added a shutdown crate which handles shutting down app-services.  The idea is that when the app goes into shutdown mode, or goes to sleep, you can call `shutdown::shutdown()` and this will interrupt all in-progress DB queries, then make the `err_if_shutdown()` start returning errors which should cause long-runner operations to start failing.  If you want to resume after sleeping you can `shutdown::restart()`.  So this is like a one-shot approach, but it also supports resuming afterwards.

This required a lot refactoring and probably more is still needed:
  - The places autocomplete code needs some refactoring and also we should implement interruption support there
  - The other components would need to be refactored to use this approach (I think this should be easier than places)
  - The iOS and Android code needs to be updated to deal with the breaking changes.  I think this should be pretty trivial.

I updated places-utils to work with this code.  To make things compile, I just ripped out all the import/export code.  The iOS/fennec code is going to go a way soon.  I wasn't sure if the other import/export code was useful, we need it then I guess that's one more coding task.  You can use places-utils to test this, by doing multiple passes and hitting `ctrl-c` in the middle to initiate shutdown/resume.  For example:

 - Execute `cargo run --example places-utils sync --nsyncs 4 --wait 1000`
 - In the middle of the first sync hit ctrl-c
 - During the 1s wait period hit ctrl-c again
 - You should see that the first sync gets interrupted, but the subsequent syncs work.

@mhammond what do you think of this approach?

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/places/src/api/places_api.rs [hunk 1]
```diff
@@ -450,32 +413,24 @@ impl PlacesApi {
...
-        // called in startup.
-        let _guard = self.sync_state.lock();
-        let conn = self.get_sync_connection()?;
-        let db = conn.lock();
-        Ok(Arc::new(db.new_interrupt_handle()))
+impl shutdown::ShutdownInterrupt for PlacesApi {
+    fn interrupt(&self) {
+        self.read_connection.interrupt_handle.interrupt();
+        self.write_connection.interrupt_handle.interrupt();
+        // TODO interrupt the sync connection, if it's open
```
Comment (bendk) [inline components/places/src/api/places_api.rs:433]: I got to a point where I realized there was already a ton of changes and I should discuss them with someone before making more.

File: components/support/shutdown/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+pub fn shutdown() {
+    IN_SHUTDOWN.store(true, Ordering::Relaxed);
+    for weak in REGISTERED_INTERRUPTS.lock().iter() {
+        if let Some(interrupt) = weak.upgrade() {
+            interrupt.interrupt()
+        }
+    }
+}
+
+/// Restart after a shutdown
```
Comment (mhammond) [inline components/support/shutdown/src/lib.rs:28]: How is this used? I wonder if we can soften expectations here, saying something like "used only for testing" or similar? Or even better, only make it available in the test cfg? Specifically, I'm thinking that to make this truly reliable we'd probably need additional protections, ensuring other threads aren't currently interrupting etc?

File: examples/places-utils/src/places-utils.rs [hunk 3]
```diff
@@ -43,127 +32,8 @@ struct DesktopItem {
     children: Vec<DesktopItem>,
 }
 
-fn convert_node(dm: DesktopItem) -> Option<BookmarkTreeNode> {
```
Comment (mhammond) [inline examples/places-utils/src/places-utils.rs:46]: I'd rather not lose this code if we can avoid it. It certainly doesn't need to be interruptable though.

File: examples/places-utils/src/places-utils.rs [hunk 4]
```diff
@@ -366,9 +204,18 @@ fn main() -> Result<()> {
...
     let api = PlacesApi::new(&db_path)?;
-    let db = api.open_connection(ConnectionType::ReadWrite)?;
     // Needed to make the get_registered_sync_engine() calls work.
     api.clone().register_with_sync_manager();
+    ctrlc::set_handler(move || {
+        if !shutdown::in_shutdown() {
+            println!("\nCTRL-C detected, enabling shutdown mode\n");
+            shutdown::shutdown();
+        } else {
+            println!("\nCTRL-C detected disabling shutdown mode\n");
```
Comment (mhammond) [inline examples/places-utils/src/places-utils.rs:214]: this seems a little suspect - if I hit ctrl+c multiple times I might end up in a frankenstate? (Not that I really care for this utility, but I'm trying to understand things)

File: components/places/src/api/places_api.rs [hunk 5]
```diff
@@ -110,15 +109,19 @@ pub struct SyncState {
...
 }
 
 /// The entry-point to the places API. This object gives access to database
 /// connections and other helpers. It enforces that only 1 write connection
 /// can exist to the database at once.
 pub struct PlacesApi {
     db_name: PathBuf,
-    write_connection: Mutex<Option<PlacesDb>>,
+    pub read_connection: Arc<PlacesConnection>,
+    pub write_connection: Arc<PlacesConnection>,
```
Comment (mhammond) [inline components/places/src/api/places_api.rs:124]: places has many read connections. Any reason you can't put interrupt support directly on `PlacesDb`, and "unregister" when dropped?

File: components/places/src/api/places_api.rs (hunk 1 above)
Comment (mhammond) [reply components/places/src/api/places_api.rs:433]: Yeah, I think we should see if we can use a lighter touch here.

**Review by mhammond — COMMENTED** [review]
In many ways I really like this. However, once I got to the semantic changes to places I became far more wary. My main idea of a one-shot shutdown was that it would allow for a lighter touch, but that seems to have failed fairly spectacularly. The fact we want to drop code from places-utils and also perform such radical surgery on the places connection semantics mean I'd probably rather just declare no interrupt/shutdown support given no one actually seems to care.

It would be great if you can work out how to stop this bleeding out and forcing us into design choices we wouldn't otherwise make.

**Pull request closed without merging**